### PR TITLE
Assert that CacheFile does not read beyond file length

### PR DESCRIPTION
**Pull request by tlrx**
In #105570 we allowed classes extending `SharedBlobCacheService` to override the method to compute the size of cached regions that can be written. This allows implementations to use the full region size whatever the length of the file/blob to cache is, and ultimately it allows to append more bytes to the last cached region of an already cached file.

While the underlying `SparseFileTracker` already makes assertions about accessing bytes within the size of the cached region, we are lacking assertions about `CacheFile` not accessing bytes beyond the end of file.

This pull request adds those assertions, in a way that classes extending `SharedBlobCacheService` can disable the assertion when needed.

### CODE REVIEW

**Review by henningandersen — COMMENTED** [review]
I have a comment on the location of the length, I am not sure it is in the right place.

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java [hunk 1]
```diff
@@ -692,13 +710,16 @@ class CacheFileRegion extends EvictableRefCounted {
 
         final RegionKey<KeyType> regionKey;
         final SparseFileTracker tracker;
+        final int regionLength;
+        final long fileLength; // used in assertions only
```
[the remark this answers is not in the document]
Comment (tlrx) [reply x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java]: I think it is kind of the same, as the file length is captured at the time the CacheFile object is created from the last known blob length.

But 6097b49fd1ae36a0b714afc461053ab7fddc5b22 might be simpler, is that what you were expected?

**Review by henningandersen — APPROVED** [review]
LGTM.

**Comment by tlrx** [comment]
Thanks Henning!

**Pull request merged**